### ISSUE TITLE
AssetManager::setup initializes in REST requests

### ISSUE THREAD

**Issue by Biont** [inpsyde/assets#8]
During REST requests the `AssetHookResolver` will conclude that we are dealing with a frontend request and return with `[ 'wp_enqueue_scripts' ]`

This can result in the unwanted execution of script configuration code ( -> in our case, we were seeing some pretty heavy logic used for `wp_localize_script` being called everywhere, even in REST requests )

However, the solution is not as easy as to add `$isRest = defined('REST_REQUEST')` to the logic since that constant is only defined *after* AssetManager bootstrap. I guess you'd need to actually inspect the `$REQUEST` route to determine a rest request.

As a workaround, we now simply do not configure the script unless we're absolutely sure we need it. But it would be nice if the assets package would remain silent when running "headless"

**Comment by Chrico** [comment]
We probably can implement some similar logic like we have in our WP App Container: https://github.com/inpsyde/wp-app-container/blob/master/src/Context.php#L69 to detect if the current request is a REST request.

**Comment by gmazzap** [comment]
Maybe is it time for the `Context` class to leave in a separate repo? We can require in WP App Container and in other places that are not necessarily WP App Container packages...

**Comment by Chrico** [comment]
Might be a good idea :-)

**Comment by gmazzap** [comment]
`isLogin()` [is there](https://github.com/inpsyde/wp-app-container/blob/master/src/Context.php#L188).

Preview and customizer are not really _contexts_ in the sense of that class.

For that class a context is something that is exclusive from other contexts. E.g. if it is frontend in not backend, and if is backend is not REST and so on (exception is `CORE` whic is a special case).

But "customizer" and "preview" for that class are sub-contexts of "backoffice".

If I would add `isCustomizer()` or `isPreview()` then why not `isPostAdminScreen()` or `isUserAdminScreen()`...

These are things that are specific of the need of this package because it deals with assets.

But still this package can benefit of the class, [`AssetHookResolver::resolve` method](https://github.com/inpsyde/assets/blob/master/src/AssetHookResolver.php#L24) could become:

```php
    public function resolve(): array
    {
        if ($this->context->isAjax() || $this->context->isRest()) {
            return [];
        }

        if ($this->context->isLogin()) {
            return [Asset::HOOK_LOGIN];
        }

        if ($this->context->isFrontoffice()) {
            return[ Asset::HOOK_FRONTEND];
        }

        if (!$this->context->isBackoffice()) {
            return [];
        }

        $hooks = is_customize_preview() ? [Asset::HOOK_CUSTOMIZER] : [];
        if (in_array($GLOBALS['pagenow'], ['post.php', 'post-new.php'], true)) {
            $hooks[] = Asset::HOOK_BLOCK_EDITOR_ASSETS;
        }
        $hooks[] = Asset::HOOK_BACKEND;

        return $hooks;
    }
```

Which is both shorter and more reliable. The only thing to consider is that global `pagenow`, just like `is_customize_preview` is nt something you can call very early. So, if `AssetManager::setup` is called e.g. on `init`, customizer and  editor assets will not work.

I can extract `Context` in a separate repo, and then send a PR here to fix this.

**Comment by Chrico** [comment]
This was implemented with #11 and is already merged. Close. :)
